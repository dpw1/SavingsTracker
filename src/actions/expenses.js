import uuid from "uuid";
import database from "../firebase/firebase";

//ADD_EXPENSE
export const addExpense = expense => ({
  type: "ADD_EXPENSE",
  expense
});

export const startAddExpense = (expenseData = {}) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    const {
      description = "",
      type = "basic-needs",
      note = "",
      amount = 0,
      createdAt = 0
    } = expenseData;

    const expense = { description, type, note, amount, createdAt };

    database
      .ref(`users/${uid}/expenses`)
      .push(expense)
      .then(ref => {
        dispatch(
          addExpense({
            id: ref.key,
            ...expense
          })
        );
      });
  };
};

export const removeExpense = ({ id } = {}) => ({
  type: "REMOVE_EXPENSE",
  id
});

export const startRemoveExpense = ({ id } = {}) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    return database
      .ref(`users/${uid}/expenses/${id}`)
      .remove()
      .then(() => {
        dispatch(removeExpense({ id }));
      });
  };
};

export const editExpense = (id, updates) => ({
  type: "EDIT_EXPENSE",
  id,
  updates
});

export const startEditExpense = (id, updates) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    return database
      .ref(`users/${uid}/expenses/${id}`)
      .update({
        ...updates
      })
      .then(() => {
        dispatch(editExpense(id, updates));
      });
  };
};

export const setExpenses = expenses => ({
  type: "SET_EXPENSES",
  expenses
});

export const startSetExpenses = () => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    return database
      .ref(`users/${uid}/expenses`)
      .once("value")
      .then(snapshot => {
        let expenses = [];

        snapshot.forEach(child => {
          expenses.push({
            id: child.key,
            ...child.val()
          });
        });

        dispatch(setExpenses(expenses));
      });
  };
};
