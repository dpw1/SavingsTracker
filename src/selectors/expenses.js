import moment from "moment";

export const getVisibleExpenses = (
  expenses,
  { text, sortBy, sortByLowToHigh, startDate, endDate }
) => {
  return expenses
    .filter(expense => {
      const createdAtMoment = moment(expense.createdAt);
      const startDateMatch = startDate
        ? startDate.isSameOrBefore(createdAtMoment, "day")
        : true;
      const endDateMatch = endDate
        ? endDate.isSameOrAfter(createdAtMoment, "day")
        : true;
      const textMatch = RegExp(text.toLowerCase()).test(
        expense.description.toLowerCase()
      );

      return startDateMatch && endDateMatch && textMatch;
    })
    .sort((a, b) => {
      const lowToHigh = sortByLowToHigh === "low-to-high" ? 1 : -1;
      if (sortBy === "date") {
        return a.createdAt < b.createdAt ? lowToHigh * 1 : lowToHigh * -1;
      } else if (sortBy === "amount") {
        return a.amount > b.amount ? lowToHigh * 1 : lowToHigh * -1;
      } else if (sortBy === "type") {
        return a.type > b.type ? lowToHigh * 1 : lowToHigh * -1;
      } else if (sortBy === "alphabetical") {
        return a.description > b.description ? lowToHigh * 1 : lowToHigh * -1;
      }
    });
};
