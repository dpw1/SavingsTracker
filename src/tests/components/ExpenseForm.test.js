import React from "react";
import { shallow } from "enzyme";
import ExpenseForm from "../../components/ExpenseForm";
import expenses from "../fixtures/expenses";

test("render ExpenseForm correctly", () => {
  const wrapper = shallow(<ExpenseForm />);
  expect(wrapper).toMatchSnapshot();
});

test("render ExpenseForm with data", () => {
  const wrapper = shallow(<ExpenseForm expense={expenses[1]} />);
  expect(wrapper).toMatchSnapshot();
});
