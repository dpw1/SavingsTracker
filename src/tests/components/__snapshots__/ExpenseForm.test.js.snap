// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render ExpenseForm correctly 1`] = `
<div
  className="expense-form"
>
  <div
    className="expense-form__message"
  />
  <form
    onSubmit={[Function]}
  >
    <input
      autoFocus={true}
      onChange={[Function]}
      placeholder="Description"
      type="text"
      value=""
    />
    <select
      id=""
      name="type"
      onChange={[Function]}
    >
      <option
        value="basic-needs"
      >
        Basic Needs
      </option>
      <option
        value="miscellaneous"
      >
        Miscellaneous
      </option>
      <option
        value="charity"
      >
        Charity
      </option>
      <option
        value="savings"
      >
        Savings
      </option>
    </select>
    <textarea
      id=""
      name="note"
      onChange={[Function]}
      placeholder="Note"
      value=""
    />
    <SingleDatePicker
      anchorDirection="left"
      customCloseIcon={null}
      customInputIcon={null}
      date={"1970-01-01T00:00:00.000Z"}
      daySize={39}
      disabled={false}
      displayFormat={[Function]}
      enableOutsideDays={false}
      firstDayOfWeek={null}
      focused={false}
      hideKeyboardShortcutsPanel={false}
      horizontalMargin={0}
      id="date"
      initialVisibleMonth={null}
      inputIconPosition="before"
      isDayBlocked={[Function]}
      isDayHighlighted={[Function]}
      isOutsideRange={[Function]}
      isRTL={false}
      keepOpenOnDateSelect={false}
      monthFormat="MMMM YYYY"
      navNext={null}
      navPrev={null}
      numberOfMonths={1}
      onClose={[Function]}
      onDateChange={[Function]}
      onFocusChange={[Function]}
      onNextMonthClick={[Function]}
      onPrevMonthClick={[Function]}
      openDirection="down"
      orientation="horizontal"
      phrases={
        Object {
          "calendarLabel": "Calendar",
          "chooseAvailableDate": [Function],
          "clearDate": "Clear Date",
          "closeDatePicker": "Close",
          "dateIsUnavailable": [Function],
          "enterKey": "Enter key",
          "escape": "Escape key",
          "hideKeyboardShortcutsPanel": "Close the shortcuts panel",
          "homeEnd": "Home and end keys",
          "jumpToNextMonth": "Move forward to switch to the next month",
          "jumpToPrevMonth": "Move backward to switch to the previous month",
          "keyboardNavigationInstructions": "Press the down arrow key to interact with the calendar and
          select a date. Press the question mark key to get the keyboard shortcuts for changing dates.",
          "keyboardShortcuts": "Keyboard Shortcuts",
          "leftArrowRightArrow": "Right and left arrow keys",
          "moveFocusByOneDay": "Move backward (left) and forward (right) by one day",
          "moveFocusByOneMonth": "Switch months",
          "moveFocusByOneWeek": "Move backward (up) and forward (down) by one week",
          "moveFocustoStartAndEndOfWeek": "Go to the first or last day of a week",
          "openThisPanel": "Open this panel",
          "pageUpPageDown": "page up and page down keys",
          "questionMark": "Question mark",
          "returnFocusToInput": "Return to the date input field",
          "selectFocusedDate": "Select the date in focus",
          "showKeyboardShortcutsPanel": "Open the keyboard shortcuts panel",
          "upArrowDownArrow": "up and down arrow keys",
        }
      }
      placeholder="Date"
      readOnly={false}
      renderCalendarInfo={null}
      renderDay={null}
      renderMonth={null}
      reopenPickerOnClearDate={false}
      required={false}
      screenReaderInputMessage=""
      showClearDate={false}
      showDefaultInputIcon={false}
      weekDayFormat="dd"
      withFullScreenPortal={false}
      withPortal={false}
    />
    <input
      name="Amount"
      onChange={[Function]}
      onKeyDown={[Function]}
      placeholder="Price"
      type="text"
      value=""
    />
    <button>
      Add Expense
    </button>
  </form>
</div>
`;

exports[`render ExpenseForm with data 1`] = `
<div
  className="expense-form"
>
  <div
    className="expense-form__message"
  />
  <form
    onSubmit={[Function]}
  >
    <input
      autoFocus={true}
      onChange={[Function]}
      placeholder="Description"
      type="text"
      value="Grocery Store"
    />
    <select
      id=""
      name="type"
      onChange={[Function]}
    >
      <option
        value="basic-needs"
      >
        Basic Needs
      </option>
      <option
        value="miscellaneous"
      >
        Miscellaneous
      </option>
      <option
        value="charity"
      >
        Charity
      </option>
      <option
        value="savings"
      >
        Savings
      </option>
    </select>
    <textarea
      id=""
      name="note"
      onChange={[Function]}
      placeholder="Note"
      value=""
    />
    <SingleDatePicker
      anchorDirection="left"
      customCloseIcon={null}
      customInputIcon={null}
      date={"1969-12-28T00:00:00.000Z"}
      daySize={39}
      disabled={false}
      displayFormat={[Function]}
      enableOutsideDays={false}
      firstDayOfWeek={null}
      focused={false}
      hideKeyboardShortcutsPanel={false}
      horizontalMargin={0}
      id="date"
      initialVisibleMonth={null}
      inputIconPosition="before"
      isDayBlocked={[Function]}
      isDayHighlighted={[Function]}
      isOutsideRange={[Function]}
      isRTL={false}
      keepOpenOnDateSelect={false}
      monthFormat="MMMM YYYY"
      navNext={null}
      navPrev={null}
      numberOfMonths={1}
      onClose={[Function]}
      onDateChange={[Function]}
      onFocusChange={[Function]}
      onNextMonthClick={[Function]}
      onPrevMonthClick={[Function]}
      openDirection="down"
      orientation="horizontal"
      phrases={
        Object {
          "calendarLabel": "Calendar",
          "chooseAvailableDate": [Function],
          "clearDate": "Clear Date",
          "closeDatePicker": "Close",
          "dateIsUnavailable": [Function],
          "enterKey": "Enter key",
          "escape": "Escape key",
          "hideKeyboardShortcutsPanel": "Close the shortcuts panel",
          "homeEnd": "Home and end keys",
          "jumpToNextMonth": "Move forward to switch to the next month",
          "jumpToPrevMonth": "Move backward to switch to the previous month",
          "keyboardNavigationInstructions": "Press the down arrow key to interact with the calendar and
          select a date. Press the question mark key to get the keyboard shortcuts for changing dates.",
          "keyboardShortcuts": "Keyboard Shortcuts",
          "leftArrowRightArrow": "Right and left arrow keys",
          "moveFocusByOneDay": "Move backward (left) and forward (right) by one day",
          "moveFocusByOneMonth": "Switch months",
          "moveFocusByOneWeek": "Move backward (up) and forward (down) by one week",
          "moveFocustoStartAndEndOfWeek": "Go to the first or last day of a week",
          "openThisPanel": "Open this panel",
          "pageUpPageDown": "page up and page down keys",
          "questionMark": "Question mark",
          "returnFocusToInput": "Return to the date input field",
          "selectFocusedDate": "Select the date in focus",
          "showKeyboardShortcutsPanel": "Open the keyboard shortcuts panel",
          "upArrowDownArrow": "up and down arrow keys",
        }
      }
      placeholder="Date"
      readOnly={false}
      renderCalendarInfo={null}
      renderDay={null}
      renderMonth={null}
      reopenPickerOnClearDate={false}
      required={false}
      screenReaderInputMessage=""
      showClearDate={false}
      showDefaultInputIcon={false}
      weekDayFormat="dd"
      withFullScreenPortal={false}
      withPortal={false}
    />
    <input
      name="Amount"
      onChange={[Function]}
      onKeyDown={[Function]}
      placeholder="Price"
      type="text"
      value="2"
    />
    <button>
      Add Expense
    </button>
  </form>
</div>
`;
